### ISSUE TITLE
GFM enhancement, do not remove _ (underscore) in words from html id tags

### ISSUE THREAD

**Issue by chadpowers** [gettalong/kramdown#267]
I have another addition to the GFM parser. For headers involving involving underscores, i.e. for
`### variable_name` Github leaves the underscore in the html id attribute, i.e. Github produces `id="variable_name"` while kramdown creates`id="variablename"`. Can you add this enhancement to the GFM parser to mimic Github's html id generation? Thanks!

**Comment by gettalong** [comment]
@chadpowers I was looking into this, do you have a live example for this? I.e. source code and resulting HTML produced by Github?

**Comment by s3ththompson** [comment]
You could perhaps just replace [`basic_generate_id`](https://github.com/gettalong/kramdown/blob/c3e6fae8b3d18d259e5ff4a2300b30bcba907140/lib/kramdown/converter/base.rb#L247) with the [implementation](https://github.com/jch/html-pipeline/blob/master/lib/html/pipeline/toc_filter.rb#L39) that GitHub uses:

```rb
PUNCTUATION_REGEXP = /[^\p{Word}\- ]/u

def basic_generate_id(str)
  gen_id = str.downcase
  gen_id.gsub!(PUNCTUATION_REGEXP, '')
  gen_id.gsub!(' ', '-')
  gen_id
end
```

I've run into this issue as well. GitHub turns `<h1>32-bit Integer</h1>` into `<h1 id="32-bit-integer">32-bit Integer</h1>`, while Kramdown yields `<h1 id="bit-integer">32-bit Integer</h1>`.

**Comment by gettalong** [comment]
@s3ththompson Yeah, the problem is that there is no spec for GFM, so we can only guess and use available live examples.

**Comment by gettalong** [comment]
@parkr Is Jekyll and/or Github Pags using the kramdown parser or kramdown's GFM parser as default? If it is the latter we should probably be careful regarding changes in the ID generation to avoid problems with existing documents.

**Comment by parkr** [comment]
@gettalong Hi! It uses the GFM parser. [Our defaults are `kramdown.input: GFM` with `kramdown.hard_wrap: false`.](https://github.com/github/pages-gem/blob/8a5b84c5c15dcf53a67d6c241591e2a725a34bd2/lib/github-pages/configuration.rb#L40-L43) Thanks for checking! 😸

**Comment by gettalong** [comment]
@parkr Thanks! What do you think: Should we make a hard change now, i.e. adopt Github's ID generation as pointed out by @s3ththompson in https://github.com/gettalong/kramdown/issues/267#issuecomment-256092013 or just change the two mentioned cases (numbers at start and underscores)?

Regarding the needed changes:

I have added some more test cases to a wiki page to see live examples - see https://github.com/gettalong/kramdown/wiki/GFM-test-case.

We will need to set the ID in the GFM parser itself since the normal automatic ID generation is done in the HTML converter.

**Comment by parkr** [comment]
@gettalong Anything to move the GFM parser closer to actual GFM is 👍 from me. /cc @benbalter @jldec if they have other thoughts.

**Comment by gettalong** [comment]
@parkr Okay, then I will update the GFM parser so that the ID generation logic from https://github.com/gettalong/kramdown/issues/267#issuecomment-256092013 is used.

**Comment by gettalong** [comment]
Since we have to do this in the parser (otherwise we would need to create a completely separate HTML converter just for GFM), the results will probably be not completely the same, ie. there *may* be some cases with different results.

**Comment by gettalong** [comment]
@chadpowers @s3ththompson @parkr I have implemented this, see commit 257ec15 - can you test this out some more? The test cases from https://github.com/gettalong/kramdown/wiki/GFM-test-case work fine.

I intend to release the next version some time next week.

**Comment by gettalong** [comment]
Will be in the release that goes out today.
